### PR TITLE
Move TagHelperBinding out of Legacy.

### PR DESCRIPTION
**Pull request by NTaylorMullen**
- Added additional properties to the class to make it more production ready.

#1092

### CODE REVIEW

File: src/Microsoft.AspNetCore.Razor.Language/TagHelperBinding.cs [hunk 1]
```diff
@@ -3,20 +3,36 @@
...
             Descriptors = _mappings.Keys;
         }
 
         public IEnumerable<TagHelperDescriptor> Descriptors { get; }
 
+        public string TagName { get; }
+
+        public string ParentTagName { get; }
+
+        public IEnumerable<KeyValuePair<string, string>> Attributes { get; }
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Razor.Language/TagHelperBinding.cs]: Should this be read only?

File: src/Microsoft.AspNetCore.Razor.Language/Legacy/TagHelperDescriptorProvider.cs [hunk 2]
```diff
@@ -97,7 +97,11 @@ public TagHelperDescriptorProvider(string tagHelperPrefix, IEnumerable<TagHelper
                 return null;
             }
 
-            var tagMappingResult = new TagHelperBinding(applicableDescriptorMappings);
+            var tagMappingResult = new TagHelperBinding(
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/TagHelperDescriptorProvider.cs]: facepalm.pgn

File: src/Microsoft.AspNetCore.Razor.Language/TagHelperBinding.cs [hunk 3]
```diff
@@ -3,20 +3,36 @@
 
 using System.Collections.Generic;
 
-namespace Microsoft.AspNetCore.Razor.Language.Legacy
+namespace Microsoft.AspNetCore.Razor.Language
 {
     public sealed class TagHelperBinding
     {
         public IReadOnlyDictionary<TagHelperDescriptor, IEnumerable<TagMatchingRule>> _mappings;
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/TagHelperBinding.cs]: fail.png

File: src/Microsoft.AspNetCore.Razor.Language/TagHelperBinding.cs [hunk 4]
```diff
@@ -3,20 +3,36 @@
...
 
         public IEnumerable<TagHelperDescriptor> Descriptors { get; }
 
+        public string TagName { get; }
+
+        public string ParentTagName { get; }
+
+        public IEnumerable<KeyValuePair<string, string>> Attributes { get; }
+
+        public string TagHelperPrefix { get; }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/TagHelperBinding.cs]: How does this get set?

**Comment by NTaylorMullen** [comment]
🆙 📅

File: test/Microsoft.AspNetCore.Razor.Language.Test/Legacy/TagHelperDescriptorProviderTest.cs [hunk 5]
```diff
@@ -10,6 +10,34 @@ namespace Microsoft.AspNetCore.Razor.Language.Legacy
...
     {
+        [Fact]
+        public void GetTagHelperBinding_ReturnsBindingWithInformation()
+        {
+            // Arrange
+            var divTagHelper = TagHelperDescriptorBuilder.Create("DivTagHelper", "SomeAssembly")
+                .TagMatchingRule(rule => rule.RequireTagName("div"))
+                .Build();
+            var descriptors = new[] { divTagHelper };
+            var attributes = new[]
```
Comment (ajaybhargavb) [inline test/Microsoft.AspNetCore.Razor.Language.Test/Legacy/TagHelperDescriptorProviderTest.cs]: nit: Rename to `expectedAttributes`

File: test/Microsoft.AspNetCore.Razor.Language.Test/Legacy/TagHelperDescriptorProviderTest.cs [hunk 6]
```diff
@@ -10,6 +10,34 @@ namespace Microsoft.AspNetCore.Razor.Language.Legacy
...
+                tagName: "th:div",
+                attributes: attributes,
+                parentTagName: "body");
+
+            // Assert
+            Assert.Equal(descriptors, bindingResult.Descriptors, TagHelperDescriptorComparer.CaseSensitive);
+            Assert.Equal("th:div", bindingResult.TagName);
+            Assert.Equal("body", bindingResult.ParentTagName);
+            Assert.Equal(attributes, bindingResult.Attributes);
+            Assert.Equal("th:", bindingResult.TagHelperPrefix);
```
Comment (ajaybhargavb) [inline test/Microsoft.AspNetCore.Razor.Language.Test/Legacy/TagHelperDescriptorProviderTest.cs]: Another assert to test `GetBoundRules()`?

**Comment by NTaylorMullen** [comment]
🆙 📅

File: src/Microsoft.AspNetCore.Razor.Language/TagHelperBinding.cs (hunk 4 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Language/TagHelperBinding.cs]: Does now.

File: src/Microsoft.AspNetCore.Razor.Language/TagHelperBinding.cs (hunk 1 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Language/TagHelperBinding.cs]: It is 😄

Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Razor.Language/TagHelperBinding.cs]: I meant exposing it as `IReadOnlyList`. But spoke offline and there is no reason to do it 👍

**Pull request merged**